docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,201 @@
+
+It ranks pods within a quality of service tier as follows:
+
+* `BestEffort` pods that consume the most of the starved resource are failed
+first
+* `Burstable` pods that consume the greatest amount of the starved resource
+relative to their request for that resource are killed first.  If no pod
+has exceeded its request, the strategy targets the largest consumer of the
+starved resource.
+* `Guaranteed` pods that consume the greatest amount of the starved resource
+relative to their request are killed first.  If no pod has exceeded its request,
+the strategy targets the largest consumer of the starved resource.
+
+### Other potential eviction strategies
+
+It is anticipated other eviction strategies may be used in the future.
+
+For example, operators may want to target `BestEffort` pods that were most
+recently scheduled on the node instead of targeting `BestEffort` pods that
+consumed the most resource.
+
+Another potential option is that operators may want to target `BestEffort` pods
+that consume the least amount of resource assuming their sum would be enough
+to reclaim enough resource on the node.  The idea being that those low resource
+pods could fit elsewhere on the cluster.
+
+Finally, operators may want to target pods based on other criteria altogether.
+For example, the time it takes for a pod to be ready and accept traffic may be
+a valid concern when looking at which pods to target for failure since many of
+those pods may get rescheduled and restart on other nodes with significant delay.
+
+Given the number of potential strategies that could be utilized, this proposal
+anticipates other strategies being defined in the future based on more specialized
+operator use cases.
+
+## Improve node stability for BestEffort pods
+
+### Feasibility checks during kubelet admission
+
+The `kubelet` will reject `BestEffort` pods if the current availability is
+below the configured eviction-threshold independent of the configured
+eviction window.
+
+The `kubelet` supports a flag to enable this behavior:
+
+* `--eviction-on-admission=true` - Do not start and fail BestEffort pods if 
+available compute resource has exceeded the eviction threshold.
+
+For example, if the `kubelet` has been configured as follows:
+
+* `--eviction-on-admission=true`
+* `--eviction-threshold=memory=256Mi`
+* `--eviction-window=5m`
+* `--eviction-monitoring-interval=1m`
+
+If the `kubelet` has less than `256Mi` of memory available on the node,
+and the node has not yet initiated eviction, the `kubelet` will immediately
+fail any incoming best effort pods.
+
+### Limit total amount of resources for BestEffort pods
+
+The `kubelet` may want to limit the total amount of resources available for
+BestEffort pods to induce their eviction even if the eviction-threshold has
+not been satisfied.
+
+The `kubelet` supports a flag to enable this behavior as follows:
+
+* `--limit-best-effort=:` A set of `ResourceName`=`ResourceQuantity` pairs that
```

Comment:
I agree that a pod of higher class should evict.

This was in response to your point 3 on this issue.

https://github.com/kubernetes/kubernetes/issues/18438

It's possible I misunderstood what you meant by a cap and if it should run separate from eviction or not require it to define a cap.